executor/src/witgen/jit/block_machine_processor.rs

```diff
@@ -0,0 +1,375 @@
+    /// Generates the JIT code for a given combination of connection and known arguments.
+    /// Fails if it cannot solve for the outputs, or if any sub-machine calls cannot be completed.
+    pub fn generate_code(
+        &self,
+        identity_id: u64,
+        known_args: BitVec,
+    ) -> Result<Vec<Effect<T, Variable>>, String> {
+        let connection = self.machine_parts.connections[&identity_id];
+        assert_eq!(connection.right.expressions.len(), known_args.len());
+
+        // Set up WitgenInference with known arguments.
+        let known_variables = known_args
+            .iter()
+            .enumerate()
+            .filter_map(|(i, is_input)| is_input.then_some(Variable::Param(i)))
+            .collect::<HashSet<_>>();
+        let fixed_evaluator = FixedEvaluatorForFixedData(self.fixed_data);
+        let mut witgen = WitgenInference::new(self.fixed_data, fixed_evaluator, known_variables);
+
+        // In the latch row, set the RHS selector to 1.
+        witgen
+            .assign(
+                &connection.right.selector,
+                self.latch_row as i32,
+                T::one().into(),
+            )
+            .unwrap();
+
+        // For each known argument, transfer the value to the expression in the connection's RHS.
+        for (index, expr) in connection.right.expressions.iter().enumerate() {
+            if known_args[index] {
+                let param_i =
+                    AffineSymbolicExpression::from_known_symbol(Variable::Param(index), None);
+                witgen.assign(expr, self.latch_row as i32, param_i).unwrap();
+            }
+        }
+
+        // Solve for the block witness.
+        // Fails if any machine call cannot be completed.
+        self.solve_block(&mut witgen)?;
+
+        // For each unknown argument, get the value from the expression in the connection's RHS.
+        // If assign() fails, it means that we weren't able to to solve for the operation's output.
+        for (index, expr) in connection.right.expressions.iter().enumerate() {
+            if !known_args[index] {
+                let param_i =
+                    AffineSymbolicExpression::from_known_symbol(Variable::Param(index), None);
+                witgen
+                    .assign(expr, self.latch_row as i32, param_i)
+                    .map_err(|_| format!("Could not solve for params[{index}]"))?;
+            }
+        }
+
+        Ok(witgen.code())
+    }
+
+    /// Repeatedly processes all identities on all rows, until no progress is made.
+    /// Fails iff there are incomplete machine calls in the latch row.
+    fn solve_block(
+        &self,
+        witgen: &mut WitgenInference<T, FixedEvaluatorForFixedData<T>>,
+    ) -> Result<(), String> {
+        let mut complete = HashSet::new();
+        for i in 0.. {
+            let mut progress = false;
+
+            // TODO: This algorithm is assuming a rectangular block shape.
+            for row in (0..self.block_size) {
```

Comment:
If we change this in the future, we should probably not use negative numbers here, because the generated code does not do the "wrap-around".

Reply:
Hmm, yeah, although I think in practice it'll be fine (that's how runtime-witgen works). It would only be a problem in cases where *both*:
- We actually use row -1 (like in `Binary`)
- The fixed columns are not fully cyclic, because the block size is not a power of 2 (Poseidon is an example)

I'm not aware of this combination in practice though.